Documentation/design-docs/assemblyloadcontext.md

```diff
@@ -0,0 +1,73 @@
+
+* For *Default LoadContext*, this override always returns *null* since *Default Context* cannot override itself.
+
+If the *Load* method override does not resolve the load, fallback to *Default LoadContext* is attempted to resolve the load incase the assembly was already loaded there. If the operating context is *Default LoadContext*, there is no fallback attempted since it has nothing to fallback to.
+
+If the *Default LoadContext* fallback also did not resolve the load (or was not applicable), the *Resolving* event is invoked against *A1's* load context. This is the last oppurtunity to attempt to resolve the assembly load. If there are no subscribers for this event, or neither resolved the load, a *FileNotFoundException* is thrown.
+
+## PInvoke Resolution
+
+*Custom LoadContext* can override the **AssemblyLoadContext.LoadUnmanagedDll** method to intercept PInvokes from within the **LoadContext** instance so that can be resolved from custom binaries. If not overridden, or if the resolution is not able to resolve the PInvoke, the default PInvoke mechanism will be used as fallback.
+
+## Constraints
+
+* **System.Private.CoreLib.dll** is only loaded once, and into the **Default LoadContext**, during the .NET Core Runtime startup as it is a logical extension of the same. It cannot be loaded into **Custom LoadContext**.
+* Currently, custom **LoadContext** cannot be unloaded once created. This is a feature we are looking into for a future release.
+* If an attempt is made to load a R2R image from the same location in multiple load context's, then precompiled code can only be used from the first image that got loaded. The subsequent images will have their code JITted. This happens because subsequent loading binaries from the same location results in OS mapping them to the same memory as the previous one was mapped to and thus, could corrupt internal state information required for use precompiled code.
+
+## Tests
+
+Tests are present [here](https://github.com/dotnet/corefx/tree/master/src/System.Runtime.Loader).
+
+## API Surface
+
+Most of the **AssemblyLoadContext** [API surface](https://github.com/dotnet/corefx/blob/master/src/System.Runtime.Loader/ref/System.Runtime.Loader.cs) is self-explanatory. Key APIs/Properties, though, are described below:
+
+### Default
+
+This property will return a reference to the *Default LoadContext*.
+
+### LoadFromAssemblyName
+
+This method can be used to load an assembly into a load context different from the load context of the currently executing assembly.
+
+## Assembly Load APIs and LoadContext
+
+As part of .NET Standard 2.0 effort, certain assembly load APIs off the **AppDomain** type, which were present in Desktop .NET Framework, have been brought back. The following maps the APIs to the load context in which they will load the assembly:
+
+* Assembly.Load - loads the assembly into the *Default LoadContext*
```

Comment:
I'm a bit concerned with how the `Assembly.Load` and `Assembly.LoadFrom` seem to be "custom ALC unfriendly", similar to `AssemblyLoadContext.Default`, as a developer can use it to write code that works fine under default ALC (the common case) but potentially break when the same code is run under a custom ALC.

As someone who is developing a plugin framework leveraging the ALC feature, I'm looking at how third party framework developers can be directed towards the "pit of success" where the code they write works seamlessly under a custom ALC.

Maybe having a "Best Practices" section about the writing ALC friendly code, or/and the API documentation for these three APIs mentioning the pitfall of using default context would be helpful. Something on the lines of _"This API always loads the assembly in the default Load Context. Consider using this [alternate API] if your code is not tied to the default load context and can run in any load context"_.

Some inputs for the proposed "Best Practices" section

1. When to override Load method vs. implement Resolve event handler for a custom ALC based on @gkhanna79 advice on this [thread](https://github.com/dotnet/coreclr/issues/8398#issuecomment-264235647).

1. `AssemblyLoadContext.Default` returns the reference to the default load context, use `AssemblyLoadContext.GetLoadContext(Assembly.GetExecutingAssembly())` if you want a reference to the current executing assembly's load context.

1. `Assembly.Load` and `Assembly.LoadFrom` always load the assembly into the default load context. If your code is not tied to the default load context and can be run in any load context, consider using `AssemblyLoadContext.Load` or `AssemblyLoadContext.LoadFromAssemblyPath` methods on the executing assembly's load context. You can get the executing assembly's load context using `AssemblyLoadContext.GetLoadContext(Assembly.GetExecutingAssembly())`.